src/board.rs

```diff
@@ -83,44 +83,81 @@ impl Board {
     pub fn move_piece(&mut self, from: Position, to: Position) -> ChessResult<()> {
         self.is_move_valid(from, to)?;
         self.squares[to.x][to.y] = self.get_piece(from).take();
-
         Ok(())
     }
 
-    fn add_moves_in_direction(&mut self, start: Position, m: Move) {
-        let mut position = start + m;
-        while Self::is_in_bounds(position).is_ok() {
-            if let Some(piece) = self.get_piece(position) {
-                // allow capturing an opponent's piece
-                if piece.get_player() != self.player {
-                    self.moves.insert((start, position));
+    fn pawn_can_double_move(&self, position: Position) -> bool {
+        match self.get_piece(position).unwrap().get_player() {
+            Player::White => position.y == 1,
+            Player::Black => position.y == 6,
+        };
+        false
+    }
+
+    fn add_pawn_advance_moves(&mut self, start: Position, player: Player) {
+        let advance_moves = Move::get_pawn_advance_moves(player, self.pawn_can_double_move(start));
+
+        for &m in advance_moves {
+            let new_position = start + m;
```

Comment:
Wait, so in the case of the pawn being on it's starting square, if there is a piece in the way of the first pawn move, it also blocks the second pawn move, even if the destination square of the second pawn move is open.

Reply:
Nice